### ISSUE TITLE
Does ttc2 reset the random seed everytime it runs? + infinite loop

### ISSUE THREAD

**Issue by vanderlindenma** [thiloklein/matchingMarkets#14]
[Weird.R.txt](https://github.com/thiloklein/matchingMarkets/files/2822643/Weird.R.txt)

EDIT: Maybe see my comments below first. 

Hi Thilo,

Me again. Sorry, I thought I was out of the bushes, but I am not. This is driving me a little crazy, but hopefully I have a better idea of what's going on now, and you might be able to help. This appears to be a really weird one, so my apologies for the long description. If it's too weird and you cannot help, I obviously understand.

I have attached a copy of the code I quote below.

I am trying to run ttc2 for a number of randomly generating profile. First I make sure the memory is empty and load the latest version of matchingMarkets:

```
### CLEAR AND LOAD ###

rm(list=ls())
library(matchingMarkets)
```

Then I generate a random profile of preferences and priorities

```
### GENERATE ###

#####
## Parameters
#####

n <- 15 # number of students
m <- 3 # number of schools
s <- 5 # number of seats per school
q <- rep(s,m) # vector of capacities
c <- 0 # degree of commonality in preferences
p <- 0 # degree of correlation between priorities and preferences
r <- 0 # degree of commonality in priorities


### Generate preferences ###

common <- runif(m)

pref <- matrix(NA, nrow = n, ncol = m)

for (i in (1:n)){
  pref[i,] <- p*common + (1-p)*runif(m) 
}

for (i in (1:n)){
  pref[i,] <- order(pref[i,])
}

### Generate priorities ###

common <- runif(n)

prior <- matrix(NA, nrow = m, ncol = n)

for (i in (1:m)){
  prior[i,] <- p*common + (1-p)*runif(n)
  for(j in (1:n)){
    prior[i,j] <- (1-r)*prior[i,j] + r*pref[j,i]
  }
}

for (i in (1:m)){
  prior[i,] <- order(prior[i,])
}
```
When I run this for the first time, I always get the following profiles

```
> prior
     [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8]
[1,]    7    2   13   12    3   15    1    5
[2,]    5    3    2    9    6   12   10    7
[3,]    5    8    2    4   15   14   12    1
     [,9] [,10] [,11] [,12] [,13] [,14] [,15]
[1,]    8    14     4     6    11    10     9
[2,]    8    15     1     4    13    14    11
[3,]    9    11    13     7     3     6    10
> pref
      [,1] [,2] [,3]
 [1,]    1    3    2
 [2,]    2    1    3
 [3,]    3    1    2
 [4,]    3    2    1
 [5,]    3    2    1
 [6,]    2    3    1
 [7,]    2    3    1
 [8,]    2    1    3
 [9,]    3    2    1
[10,]    1    2    3
[11,]    3    1    2
[12,]    3    1    2
[13,]    3    2    1
[14,]    3    1    2
[15,]    2    3    1
```
which is what you would expect if R always starts from the same random seed. 

(I guess? Note that I am not setting the random seed myself before doing all this. So maybe the fact that I always get the same random outcome indicates something fishy? Or does R by default always start with the same random seed on any given machine?).

As would be expected, if I run ### GENERATE ### again, I get a different profile, in my case

```
> prior
     [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8]
[1,]   13   15    1    5    6   11   10    2
[2,]   14    8    9   12   11    7    6    4
[3,]    9    5    1   10    2   11   12   15
     [,9] [,10] [,11] [,12] [,13] [,14] [,15]
[1,]   14     8    12     4     3     9     7
[2,]   10     3     2    13    15     5     1
[3,]    8    13     7     3    14     4     6
> pref
      [,1] [,2] [,3]
 [1,]    2    3    1
 [2,]    2    1    3
 [3,]    3    1    2
 [4,]    1    3    2
 [5,]    3    2    1
 [6,]    2    3    1
 [7,]    2    3    1
 [8,]    1    2    3
 [9,]    3    2    1
[10,]    2    1    3
[11,]    2    3    1
[12,]    1    3    2
[13,]    3    2    1
[14,]    3    1    2
[15,]    2    1    3
```
A first weird thing is that, if I now run ttc2, the functions gets stuck into an infinite loop. Below is the ttc2 that runs forever if I execute it after having ran ### GENERATE ### a second time and obtained the second profile.

```
ttc.o <- ttc2(
  s.prefs = t(pref), 
  c.prefs = t(prior), 
  nSlots = q
)
```
In contrast, if I run the above code after a *single* run of ### GENERATE ### (i.e., supposedly with the first profile), the function quickly completes and outputs:

```
> ttc.o
   ind obj
1    1   1
2    2   2
3    3   3
4    4   3
5    5   3
6    6   2
7    7   2
8    8   2
9    9   3
10  10   1
11  11   1
12  12   1
13  13   1
14  14   3
15  15   2
```
Now, here is where it gets even weirder. 

Suppose ttc2 got stuck in an infinite loop after I try to run it following two executions of ### GENERATE ### and I kill the execution (hit the red "stop" button in RStudio). Then, if I run the whole code again, ttc2 produces an output. But somehow, having killed an the infinite execution of ttc2 seems to have re-initialized the seed, as I once again get my initial profile from the random generation. 

Here is the output when I re-run after having killed:

```
> ### Generate preferences ###
> 
> common <- runif(m)
> 
> pref <- matrix(NA, nrow = n, ncol = m)
> 
> for (i in (1:n)){
+   pref[i,] <- p*common + (1-p)*runif(m) 
+ }
> 
> for (i in (1:n)){
+   pref[i,] <- order(pref[i,])
+ }
> 
> ### Generate priorities ###
> 
> common <- runif(n)
> 
> prior <- matrix(NA, nrow = m, ncol = n)
> 
> for (i in (1:m)){
+   prior[i,] <- p*common + (1-p)*runif(n)
+   for(j in (1:n)){
+     prior[i,j] <- (1-r)*prior[i,j] + r*pref[j,i]
+   }
+ }
> 
> for (i in (1:m)){
+   prior[i,] <- order(prior[i,])
+ }
> 
> #### Print ###
> 
> prior
     [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8]
[1,]    7    2   13   12    3   15    1    5
[2,]    5    3    2    9    6   12   10    7
[3,]    5    8    2    4   15   14   12    1
     [,9] [,10] [,11] [,12] [,13] [,14] [,15]
[1,]    8    14     4     6    11    10     9
[2,]    8    15     1     4    13    14    11
[3,]    9    11    13     7     3     6    10
> pref
      [,1] [,2] [,3]
 [1,]    1    3    2
 [2,]    2    1    3
 [3,]    3    1    2
 [4,]    3    2    1
 [5,]    3    2    1
 [6,]    2    3    1
 [7,]    2    3    1
 [8,]    2    1    3
 [9,]    3    2    1
[10,]    1    2    3
[11,]    3    1    2
[12,]    3    1    2
[13,]    3    2    1
[14,]    3    1    2
[15,]    2    3    1
> 
> ### Run TTC2 ###
> 
> ttc.o <- ttc2(
+   s.prefs = t(pref), 
+   c.prefs = t(prior), 
+   nSlots = q
+ )
> #### Print ###
> 
> ttc.o
   ind obj
1    1   1
2    2   2
3    3   3
4    4   3
5    5   3
6    6   2
7    7   2
8    8   2
9    9   3
10  10   1
11  11   1
12  12   1
13  13   1
14  14   3
15  15   2
```
A similar thing happen if I just run the whole code over an over, i.e., I always get the same random profile, suggesting that running ttc2 somehow resets the random seed ???

In summary:

- ttc2 runs well only if I use the profile generated by the first run of my random process.
- If I try to run ttc2 with another profile generated from a second run of the random process, ttc2 gets into an infinite loop.
- And finally, after any run of ttc2, whether successful or aborted, my random profile generator always spits out the same random profile as if running ttc2 was resetting the random seed ???

At this point, I am clueless. What the hell might be going on? Any help would be appreciated.

Thanks in advance,

**Comment by vanderlindenma** [comment]
The code below (using the example from ttc2's documentation) still does not explain why ttc2 runs into an infinite loop after two runs of ### GENERATE# above.

However, it suggests that running ttc2 does indeed reset the random seed:

```
> s.prefs <- matrix(c(2,1,3,4,1,2,3,4,3,2,1,4,3,4,1,2,1,3,4,2,4,1,2,3,1,2,3,4,1,2,4,3),byrow = FALSE, ncol = 8); s.prefs
     [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8]
[1,]    2    1    3    3    1    4    1    1
[2,]    1    2    2    4    3    1    2    2
[3,]    3    3    1    1    4    2    3    4
[4,]    4    4    4    2    2    3    4    3
> c.prefs <- matrix(c(1,2,3,4,5,6,7,8,3,5,4,8,7,2,1,6,5,3,1,7,2,8,6,4,6,8,7,4,2,3,5,1),byrow = FALSE, ncol = 4); c.prefs
     [,1] [,2] [,3] [,4]
[1,]    1    3    5    6
[2,]    2    5    3    8
[3,]    3    4    1    7
[4,]    4    8    7    4
[5,]    5    7    2    2
[6,]    6    2    8    3
[7,]    7    1    6    5
[8,]    8    6    4    1
> nSlots <- c(2,2,3,3)
> 
> runif(1)
[1] 0.899825
> ttc2(s.prefs = s.prefs, c.prefs = c.prefs, nSlots = nSlots)
  ind obj
1   1   2
2   2   1
3   3   3
4   4   3
5   5   1
6   6   4
7   7   2
8   8   4
> runif(1)
[1] 0.551435
> ttc2(s.prefs = s.prefs, c.prefs = c.prefs, nSlots = nSlots)
  ind obj
1   1   2
2   2   1
3   3   3
4   4   3
5   5   1
6   6   4
7   7   2
8   8   4
> runif(1)
[1] 0.551435
> ttc2(s.prefs = s.prefs, c.prefs = c.prefs, nSlots = nSlots)
  ind obj
1   1   2
2   2   1
3   3   3
4   4   3
5   5   1
6   6   4
7   7   2
8   8   4
> runif(1)
[1] 0.551435
> ttc2(s.prefs = s.prefs, c.prefs = c.prefs, nSlots = nSlots)
  ind obj
1   1   2
2   2   1
3   3   3
4   4   3
5   5   1
6   6   4
7   7   2
8   8   4
```
Is that intended?

**Comment by vanderlindenma** [comment]
Ok, yes, of course:

```
ttc2 <- function( nStudents = ncol(s.prefs), nColleges = ncol(c.prefs), s.prefs = NULL,  c.prefs = NULL,nSlots = NULL, priority = NULL , seed = 123, full_return = FALSE){
  
  ## To Do
  ## Slots in find_cycle aktualisieren?
  
  # Check priority
  if(missing(priority)){
    set.seed(seed)
    priority <- sample(1:nStudents)   # Assign random priority ordering if none is given
  }
```

**Comment by vanderlindenma** [comment]
Actually, I am not clear on what "priority" does to ttc2? It seems to be related to rsd rather than ttc2. Does ttc2 override c.prefs and run a serial dictatorship instead when provided with a priority order?

**Comment by AlexSauer** [comment]
Hello,

thank you very much for your comments and the time you invested in the problem!
Indeed, there is a bug in the code and we will fix it in the next update.

If you do not want to wait until then, you can fix the problem yourself:
- If you call ttc2 and set seed = NULL, i.e. ttc2(..., seed = NULL), than no seed should be set.
- Concerning the infinite loop, until the update the best thing you can do is the following: Copy the code of the ttc2 function to a local file on your machine and change in line 165
'   for(j in 1:dim(s.prefs)[1]){ ' to  ' for(j in 1:dim(s.prefs)[2]){ '
then load the file after you loaded the matchingMarkets-package. 
This fixed the problem for us, but if you still run into an infinite loop please don't hesitate to give us more feedback.
- The description of the priority parameter might be misleading: This has nothing to do with the the preferences of either side. The algorithm searches for cycles and priority only gives the ranking of the students in the search for cycles, i.e. the student with the highest priority is the first who is allowed to 'point' to some school.

We hope we could help you and thanks again!
Alex

**Comment by thiloklein** [comment]
This has been fixed with the latest commit on GitHub and just been pushed to CRAN.

**Comment by vanderlindenma** [comment]
Thanks a lot Thilo and Alex. I will return to that project at some point and try to run a similar code again. I'll follow up if I run into any issue.
